garden_ai/utils/_meta.py

```diff
@@ -0,0 +1,144 @@
+
+    Defines other helper functions locally to prevent polluting the __main__
+    namespace (see: `make_func_to_serialize`).
+
+    Any future "pipeline middleware" (like `inject_env_kwarg`) should also be
+    defined locally inside this function.
+    """
+
+    def compose(*functions):
+        func, *funcs = functions
+
+        def inner(*args, **kwargs):
+            result = func(*args, **kwargs)
+            for f in funcs:
+                result = f(result)
+            return result
+
+        return inner
+
+    def inject_env_kwarg(func):
+        """
+        Helper: modify a function so that it will accept an ``_env_vars`` keyword argument.
+
+        This can be used to dynamically set environment variables before executing the
+        original function, particularly useful if the function is executing remotely.
+        """
+
+        def inner(*args, _env_vars=None, **kwargs):
+            if _env_vars:
+                import os
+
+                for k, v in _env_vars.items():
+                    os.environ[k] = v
+            return func(*args, **kwargs)
+
+        return inner
+
+    return inject_env_kwarg(compose(*functions))
+
+
+def exec_getsource(source, globals=None, locals=None):
+    """
+    helper: same as built in exec, but in a way that inspect.getsource can
+    still extract source code (i.e. for steps we parsed from a pipeline.py)
+    """
+    # https://stackoverflow.com/a/69668959
+    getlines = linecache.getlines
+
+    def monkey_patch(filename, module_globals=None):
+        if filename == "<string>":
+            return source.splitlines(keepends=True)
+        else:
+            return getlines(filename, module_globals)
+
+    linecache.getlines = monkey_patch
+
+    try:
+        exec(source, globals, locals)
+        # you can now use inspect.getsource() on the result of exec() here
+
+    finally:
+        linecache.getlines = getlines
+
+
+def _load_pipeline_from_python_file(python_file):
+    import __main__
+    from garden_ai import Pipeline
+
+    with open(python_file, "r") as file:
+        pipeline_code = file.read()
+
+    # dynamically create a class that we're using exclusively for the sake of a
+    # simpler namespace than a ModuleType object.
+    #
+    # This is necessary because user code also imports garden_ai, and the more
+    # robust importlib equivalent goes too far for our needs -- we want `import
+    # garden_ai` in the user's code to refer to the already-imported garden_ai,
+    # but something about `exec_module`'s internals does a "true" re-import of
+    # garden_ai, with the unfortunate side effect of causing dill to think it
+    # needs to trace/serialize far more than it actually should.  (i.e.
+    # basically everything, including unserializable objects like the CLI's rich
+    # console).
+    code_str = f"""
+class _USER_PIPELINE_MODULE:
+    {textwrap.indent(pipeline_code, '    ')}
```

Comment:
Does this make the very first line double indented? (because there are 4 spaces before the {} block)

Like it's clear that this works because you've tested this. Just a little confused

Reply:
oh great catch, I'm realizing that everything I tested it against started with an `#!/usr/bin/env python3` so it didn't cause problems but it easily could, I'll play around with it to sanity check